### PR TITLE
Remove g_rpc_node global

### PR DESCRIPTION
**Pull request by ryanofsky**
This PR removes the `g_rpc_node` global, to get same benefits we see removing other globals and make RPC code more testable, modular, and reusable.

This uses a hybrid of the approaches suggested in #17548. Instead of using `std::any`, which isn't available in c++11, or `void*`, which isn't type safe, it uses a small new `util::Ref` helper class, which acts like a simplified `std::any` that only holds references, not values.

Motivation for writing this was to provide an simpler alternative to #18647 by Harris Brakmić (brakmic) which avoids some shortcomings of that PR (https://github.com/bitcoin/bitcoin/pull/18647#issuecomment-617878826)

### CODE REVIEW

**Comment by MarcoFalke** [comment]
> Instead [...] void*, which isn't type safe, it uses a small new util::Ref 

Why not `NodeContext*`? That might seem like a layer violation reading the source code, but practically it does the same and the diff would be smaller.

**Comment by ryanofsky** [comment]
> > Instead [...] void*, which isn't type safe, it uses a small new util::Ref
> 
> Why not `NodeContext*`? That might seem like a layer violation reading the source code, but practically it does the same and the diff would be smaller.

This would allow only allow dropping the first commit. The later commits would look basically the same but with `s/util::Ref/NodeContext/`.

The practical reason I don't want to do this is I want to the be able to use the RPC server outside the node process. E.g. start a multiprocess wallet process and have it listen for wallet RPC requests. Maybe add more offline wallet support and be able to have wallet-tool respond to RPCs like `listtransactions`. In these cases, wallet code can't link against `src/node/context.cpp` and `NodeContext` is useless for passing wallet state (aside from just looking strange).

Also, RPC code in rpc/server and httprpc is mostly pretty generic and not bitcoin specific and personally I think it would be a shame to see bitcoin types leak into it.

**Comment by promag** [comment]
From https://github.com/bitcoin/bitcoin/pull/18647#issuecomment-614059391

> Wallet RPC methods shouldn't have access to NodeContext. At some point we should have WalletContext struct that will hold things like vpwallets and let us get rid of wallet globals, and it would make sense for wallet RPC methods to have access to this. That is why suggestion from #17548 was to just add a new member to JSONRPCRequest that could hold a generic context, not something tied to the wallet or node. Other advantage of using JSONRPCRequest is that you dont' have to change RPC method signatures.

How would you implement this? Make context a `map<string, any>`?

**Comment by practicalswift** [comment]
Concept ACK

Very happy to see `g_rpc_node` go!

Thanks for the great architectural work you are doing.

**Comment by ryanofsky** [comment]
> How would you implement this? Make context a `map<string, any>`? And do `context["node"].Has<NodeContext>()`?

No need for multiple contexts. `interfaces::WalletClientImpl` can just pass an different context to wallet RPC methods. Passing an `interfaces::Chain` reference would be sufficient enough for `loadwallet` / `createwallet`. But I actually want to introduce a `WalletContext` struct in `src/wallet/context.h` analagous to the `NodeContext` struct in `src/node/context.h` which would hold things like the `Chain` interface pointer and `vpwallets` vector.

**Comment by ryanofsky** [comment]
Actually @promag, it is a little clumsier than I thought. I should change this PR to pass context to rpc method register calls instead of the StartHTTPRPC call to make this more future-proof (should still be an improvement as is)

File: src/rpc/blockchain.h [hunk 1]
```diff
@@ -47,11 +50,7 @@ UniValue blockheaderToJSON(const CBlockIndex* tip, const CBlockIndex* blockindex
 /** Used by getblockstats to get feerates at different percentiles by weight  */
 void CalculatePercentilesByWeight(CAmount result[NUM_GETBLOCKSTATS_PERCENTILES], std::vector<std::pair<CAmount, int64_t>>& scores, int64_t total_weight);
 
-//! Pointer to node state that needs to be declared as a global to be accessible
-//! RPC methods. Due to limitations of the RPC framework, there's currently no
-//! direct way to pass in state to RPC methods without globals.
-extern NodeContext* g_rpc_node;
-
-CTxMemPool& EnsureMemPool();
+NodeContext& EnsureNodeContext(const util::Ref& context);
```
Comment (promag) [inline src/rpc/blockchain.h:53]: Have you considered `EnsureNodeContext(const JSONRPCRequest& req)` instead? Same for `EnsureMemPool`. Could avoid forward declaration above.

**Review by promag — COMMENTED** [review]
Concept ACK.

File: src/rest.cpp [hunk 2]
```diff
@@ -610,7 +610,7 @@ static bool rest_blockhash_by_height(const util::Ref& context, HTTPRequest* req,
     std::string height_str;
     const RetFormat rf = ParseDataFormat(height_str, str_uri_part);
 
-    int32_t blockheight;
+    int32_t blockheight = 0;
```
Comment (MarcoFalke) [inline src/rest.cpp:613]: Heh, nice find. Maybe split this up as a separate pull request, since it seems unrelated to the changes here?

Comment (ryanofsky) [reply src/rest.cpp:613]: re: https://github.com/bitcoin/bitcoin/pull/18740#discussion_r414772851

> Maybe split this up as a separate pull request, since it seems unrelated to the changes here?

Yes it should be basically be unrelated. I created https://github.com/bitcoin/bitcoin/pull/18785, but am also keeping the commit here for now, since it does seem to be needed to avoid travis failing with the other changes here https://travis-ci.org/github/bitcoin/bitcoin/jobs/678453061#L7157

File: src/rpc/blockchain.h (hunk 1 above)
Comment (ryanofsky) [reply src/rpc/blockchain.h:53]: re: https://github.com/bitcoin/bitcoin/pull/18740#discussion_r413428813

> Have you considered `EnsureNodeContext(const JSONRPCRequest& req)` instead? Same for `EnsureMemPool`. Could avoid forward declaration above.

Probably considered it, but I don't like this style (along lines of c++ koan https://www.youtube.com/watch?v=Zx_Tjp9WIII&t=15m07s)

**Review by ryanofsky — COMMENTED** [review]
Rebased 0708140ceee683e4d6288f4ca6616118e69218dd -> 0d84b1a3a7238ca171e7494904dc0c221b91edc2 ([`pr/frpc.1`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.1) -> [`pr/frpc.2`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/frpc.1-rebase..pr/frpc.2)) to match pr base, also fixed ref clear method
Updated 0d84b1a3a7238ca171e7494904dc0c221b91edc2 -> fe333744b72c11030e5a06e904ba44cea167117b ([`pr/frpc.2`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.2) -> [`pr/frpc.3`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/frpc.2..pr/frpc.3)) adding ref.h to makefile to fix travis errors https://travis-ci.org/github/bitcoin/bitcoin/builds/678246736
Rebased fe333744b72c11030e5a06e904ba44cea167117b -> 5f260284e059504cc6f93dc575054e526470f181 ([`pr/frpc.3`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.3) -> [`pr/frpc.4`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/frpc.3-rebase..pr/frpc.4)) with scope fix and rebase due to conflict with #18671
Added 1 commits 5f260284e059504cc6f93dc575054e526470f181 -> 41d5d651594c6c939add7a58b7e30c97dccdf24a ([`pr/frpc.4`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.4) -> [`pr/frpc.5`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/frpc.4...pr/frpc.5)) to fix valgrind false positive https://travis-ci.org/github/bitcoin/bitcoin/jobs/678453061#L7157

**Comment by ryanofsky** [comment]
Rebased 41d5d651594c6c939add7a58b7e30c97dccdf24a -> 02aa19a8b1a2e7e6b2293ae8512f7c606ca58345 ([`pr/frpc.5`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.5) -> [`pr/frpc.6`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/frpc.5-rebase..pr/frpc.6)) due to conflict with #18785

**Comment by brakmic** [comment]
ACK 02aa19a8b1a2e7e6b2293ae8512f7c606ca58345

Built, run and done some manual RPC testing on macOS Catalina 10.15.4

**Comment by ryanofsky** [comment]
Rebased 02aa19a8b1a2e7e6b2293ae8512f7c606ca58345 -> 1be6bad7c044981faea55c76730a07eb95c2fd9f ([`pr/frpc.6`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.6) -> [`pr/frpc.7`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/frpc.6-rebase..pr/frpc.7)) due to conflict with #18699

**Comment by ryanofsky** [comment]
Travis error appears to be unrelated "`mkdir: cannot create directory ‘/home/travis/build/bitcoin/bitcoin/ci/scratch’: No space left on device
`" in s390x native BE build https://travis-ci.org/github/bitcoin/bitcoin/jobs/683061500#L336

**Comment by ryanofsky** [comment]
Rebased 1be6bad7c044981faea55c76730a07eb95c2fd9f -> c13af197ce473769303c3bf1bcb38658c5f8de41 ([`pr/frpc.7`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.7) -> [`pr/frpc.8`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/frpc.7-rebase..pr/frpc.8)) due to conflict with #16224

**Review by promag — COMMENTED** [review]
Code review ACK c13af197ce473769303c3bf1bcb38658c5f8de41.

**Comment by ajtowns** [comment]
> I should change this PR to pass context to rpc method register calls

I found the `std::any` versus explicit `NodeContext` tradeoff confusing, but I think this makes sense of it. With a node/wallet/mempool/whatever context passed to the rpc register calls, this means

 * the RPC infra (server, json request) doesn't need to know about bitcoin structures at all
 * the RPC command can just grab its needed context directly, and there doesn't need to be a struct with pointers to all the different parts that might or might not be available, which will maybe make it easier to split the different contexts into separate programs one day
 * the RPC server can give an error immediately for RPC commands where the context isn't available (I guess it'd be "command not found" since it wouldn't have been registered in the first place)

So, Approach ACK. Planning on updating this pr with the register changes, or doing that later?

**Comment by ryanofsky** [comment]
> So, Approach ACK. Planning on updating this pr with the register changes, or doing that later?

Was planning on doing it in a followup PR removing g_rpc_chain

File: src/test/ref_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    util::Ref ref;
+    BOOST_CHECK(!ref.Has<int>());
+    BOOST_CHECK_THROW(ref.Get<int>(), NonFatalCheckError);
+    int value = 5;
+    ref.Set(value);
+    BOOST_CHECK(ref.Has<int>());
+    BOOST_CHECK_EQUAL(ref.Get<int>(), 5);
+    ++ref.Get<int>();
+    BOOST_CHECK_EQUAL(ref.Get<int>(), 6);
+    BOOST_CHECK(!ref.Has<bool>());
```
Comment (MarcoFalke) [inline src/test/ref_tests.cpp:22]: in the commit that adds ref: Could check that each is an alias for the other? Right now ref could actually be a copy and the test wouldn't fail.

```diff
diff --git a/src/test/ref_tests.cpp b/src/test/ref_tests.cpp
index ef96a0042a..0ec0799fbc 100644
--- a/src/test/ref_tests.cpp
+++ b/src/test/ref_tests.cpp
@@ -19,6 +19,10 @@ BOOST_AUTO_TEST_CASE(ref_test)
     BOOST_CHECK_EQUAL(ref.Get<int>(), 5);
     ++ref.Get<int>();
     BOOST_CHECK_EQUAL(ref.Get<int>(), 6);
+    BOOST_CHECK_EQUAL(value, 6);
+    ++value;
+    BOOST_CHECK_EQUAL(value, 7);
+    BOOST_CHECK_EQUAL(ref.Get<int>(), 7);
     BOOST_CHECK(!ref.Has<bool>());
     BOOST_CHECK_THROW(ref.Get<bool>(), NonFatalCheckError);
     ref.Clear();

**Review by maflcko — APPROVED** [review]
ACK c13af197ce473769303c3bf1bcb38658c5f8de41 🖋

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK c13af197ce473769303c3bf1bcb38658c5f8de41 🖋
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUinRwwAkHWlUKKhw3GzOtBTqDhsNMOPj4xwTRNK2IwJujWN/z7nf7bqQLtPBEGw
7/wtrfXP8azfrGTqBmGTEqRxYjH0HiB8q0qhZvCW+oKGOvUIN1SUPLf9Dumj8oFd
HdMIo6IVJmj9weolegQxWNJ7z55E3M7ZblMnKO9pYZqGWjs/Mj2il+Aza0/UjYN1
HHQX0E8CqZSM08zDJelpkg9qKp0TacEvbJjb2SPirhd9YvxeYIDRhX6Axfdw+rJR
0ujuMTrXuJTJ2fkLhZa6mRzSVDEB/NLqgVQKou1f18epfljFq6xi/1+gFYnro21A
Jy24DqiB5Jq2kTvox2sgLa3j66E9xdiOh5OKYEi67BjENCZC2lqpKybi0p2P1yP5
S3MRaFiU1xjShXf8lyV8EUpEZDlj3bZO+zP0x3ox6qxu6vGQzOf+lbUMNQKgyOTJ
wnBVgW05G7qUGZhUZGoKWf7w2MII/IV6nNWKg/CqAO8g7zjU0rOqlwPhIcYmFQ7J
bRC2QBGP
=68aI
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `29eb95433107a5581e173c806df9747dec6d545006ccc3f9f225e8cf38bbd914  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010829eb95433107a5581e173c806df9747dec6d545006ccc3f9f225e8cf38bbd914f0103d0a8e697c381106773d84875aa7b18b08fff0100a03feff4e0713db37ea6d188df1afbd08f1045ec3f8b3f008905bc17b61b5d8040083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0104de02b6a03184f9a762f533399e50f7a08f1045ec3f8b3f00887743359e6754f200083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01070bf1bc2f82791e05a9d503361a5f45108f1045ec3f8b2f00800f10e24b7c7c9800083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010db207538aeb6a99f33adab12acd6520008f1045ec3f8b2f008ac3db337eb46d9f40083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Review by ryanofsky — COMMENTED** [review]
Thanks for the review! Updated with suggested changes

Updated c13af197ce473769303c3bf1bcb38658c5f8de41 -> b3f7f375efb9a9ca9a7a4f2caf41fe3df2262520 ([`pr/frpc.8`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.8) -> [`pr/frpc.9`](https://github.com/ryanofsky/bitcoin/commits/pr/frpc.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/frpc.8..pr/frpc.9))

File: src/rpc/request.h [hunk 4]
```diff
@@ -34,8 +38,9 @@ class JSONRPCRequest
     std::string URI;
     std::string authUser;
     std::string peerAddr;
+    const util::Ref& context;
 
-    JSONRPCRequest() : id(NullUniValue), params(NullUniValue), fHelp(false) {}
+    JSONRPCRequest(const util::Ref& context) : id(NullUniValue), params(NullUniValue), fHelp(false), context(context) {}
```
Comment (ajtowns) [inline src/rpc/request.h:43]: `Ref& _context` and `context(_context)`? We usually use different names for arguments and members, don't we?

**Review by ajtowns — COMMENTED** [review]
ACK b3f7f375efb9a9ca9a7a4f2caf41fe3df2262520

File: src/util/ref.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+#include <util/check.h>
+
+#include <typeindex>
+
+namespace util {
+
+/**
+ * Type-safe dynamic reference.
+ *
+ * This implements a small subset of the functionality in C++17's std::any
```
[the remark this answers is not in the document]
Comment (ryanofsky) [reply src/util/ref.h:17]: Thanks for the review. std::any can hold pointers and reference wrappers, so there is no problem here. I'm going to mark the style comments resolved because I disagree with them and don't want to create extended off-topic conversations. But if there's anything that should be reconsidered, definitely do follow up. Review suggestions tend to be most useful when they are specific and say something about the motivation behind the suggestion. That way, if there's disagreement about the suggestion, something else can be done to address whatever the motivation for it was.

File: src/rpc/request.h (hunk 4 above)
Comment (MarcoFalke) [reply src/rpc/request.h:43]: This will cause -Wshadow warnings, but luckily we don't have them enabled anymore

File: src/rpc/misc.cpp [hunk 6]
```diff
@@ -398,9 +399,10 @@ static UniValue mockscheduler(const JSONRPCRequest& request)
     }
 
     // protect against null pointer dereference
-    CHECK_NONFATAL(g_rpc_node);
-    CHECK_NONFATAL(g_rpc_node->scheduler);
-    g_rpc_node->scheduler->MockForward(std::chrono::seconds(delta_seconds));
+    CHECK_NONFATAL(request.context.Has<NodeContext>());
+    NodeContext& node = request.context.Get<NodeContext>();
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/rpc/misc.cpp:403]: I think russ tries to keep the previous structure of the code, but I agree this can be changed in a follow-up

**Comment by MarcoFalke** [comment]
re-ACK b3f7f375ef, only change is adding back const and more tests 🚾

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK b3f7f375ef, only change is adding back const and more tests 🚾
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjEBAv/RaWaCyOsjaMttBcKiwYoYmFFtu0dDqcwQ8icmd/JotFyemomPv1LSpOs
+rmy/8gpfyM875zlzSJ/jg0ELlQm7LS7RQecVSSYCx27J8U6olXBJILuPx4T8yJs
DY3DABlfI1lQv7rffyZu0mxRYv8b4sGh+80ZTL4t3WOuvsdnS+PsqiK+BJiG571k
9Oeg4BQxLaYKR5Q80n74ATb4tv2eogpYJ8JmKRIOZ3DWOtFsSot2WCEmaxOJc0If
RJqEGeJ/sL8+hM3lEZm4HYI/eV0/WpezgnpX8pn7kvcpnS4NcElPSzXiFYpYjKYe
T4dSFvRg5rZaqJ61PCHBjFk5mMV0sHtRQ4mbGiAiCPvhpoVGbbEq+3YY2q5+ZXs0
E2PvUzHz6wqiK7ZmtpmM3CiuA8cu3MgujgGJF5ou5ZW1HIGjne3brc2lKnl7SfNO
m5TghZ/FY8JiBnk/Wwzp7WSf3Zf9Q7SD6sqTy9/PKQEquWSNQNxsc/hziYfSpjYF
Z1obJ5Pr
=nEIH
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `4de1c5a739df99abe18f469939c72f67f85aab41eb67e464e5ba55c4a6d9c8a4  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401084de1c5a739df99abe18f469939c72f67f85aab41eb67e464e5ba55c4a6d9c8a4f010696d8c32460d8701316b0db60bb69c3b08fff01002f2162b03a7279ca112622ec96994c008f1045ec65d74f0080dcca0560482ff350083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010398b7c9f2fca286955278169e74129c908f0209141d77ac03eba1fca57ff25a347ef81de65f5bb5ac0c92ba355bc54f0abf7cc08f1045ec65d75f00872e0a437306e63b50083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0103d1e70b1e1a7b8c9470a1dd6acbe1aed08f0209211751cf40a4a8941e3aa0b028a0e439c1369919384fd523e4354ec04b5cc2808f1045ec65d75f0085fd9693165be35170083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df01098693f06ba7d21d4eeab2e7d3b8888f008f1045ec65d74f00871e6435f42fcb3db0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

File: src/rpc/request.h (hunk 4 above)
Comment (MarcoFalke) [reply src/rpc/request.h:43]: This can be fixed up in the follow-up that switches to C++17

**Comment by hebasto** [comment]
@ryanofsky 
> This uses a hybrid of the approaches suggested in #17548. Instead of using `std::any`, which isn't available in c++11, or `void*`, which isn't type safe, it uses a small new `util::Ref` helper class, which acts like a simplified `std::any` that only holds references, not values.

The `std::any` requires that a type of a contained object [must](https://en.cppreference.com/w/cpp/utility/any/any) be CopyConstructible.
But `std::is_copy_constructible<NodeContext>::value == false`.

So, even with C++17, there are no simple ways to drop `util::Ref` in favor of `std::any`, unfortunately.

Or did I miss something?

**Comment by fanquake** [comment]
> So, even with C++17, there are no simple ways to drop util::Ref in favor of std::any, unfortunately.

It's being dropped for `std::any` in #21366.
